frc: f3-augmented-snapshot (#1169)
filecoin-project/FIPs: FRCs/frc-f3-augmented-snapshot.md

---
fip: "####"
title: "Augment Filecoin Snapshot with F3 data"
author: "Hailong Mu (@hanabi1224)"
discussions-to: https://github.com/filecoin-project/go-f3/issues/480
status: Draft
type: "FRC"
created: 2025-06-25
---

> # FRC-####: Augment Filecoin Snapshot with F3 data

Reviewer: @rvagg : this is definitely something we're going to do and ship.  Any concern with assigning this FRC-0108?

> # FRC-####: Augment Filecoin Snapshot with F3 data

Reviewer: ```suggestion
# FRC-####: Filecoin Snapshot Format
```

We propose extending the Filecoin CAR snapshot with an F3 snapshot as a raw data block, and changing CAR roots to be a CID that points to a CBOR-encoded Filecoin snapshot header struct.

## Motivation
<!--The motivation is critical for FIPs that want to change the Filecoin protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the FIP solves. FIP submissions without sufficient motivation may be rejected outright.-->

The time cost and the network bandwidth usage for a new Filecoin node to catch up with all F3 finality certificates grow over time, which delays the readiness of the F3-aware V2 RPC APIs. By embedding an F3 snapshot into the current Filecoin CAR snapshot, both can be vastly reduced at the cost of a slightly increased Filecoin CAR snapshot size.

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any current Filecoin implementations. -->

We propose the blow changes to the Filecoin CAR snapshot format.

> - Change CAR roots to be a CID that points to a CBOR-encoded [`SnapshotMetadata`](#snapshotmetadata) struct that is stored as the first data block in the CAR.

Reviewer 2: Not technically wrong since it's plural, `roots` and can be an array. A pedantic edit could be:

```suggestion
- Change CAR header's `roots` array to contain a single CID that points to a CBOR-encoded [`SnapshotMetadata`](#snapshotmetadata) struct that is stored as the first data block in the CAR.
```

- Store the raw [`F3Snapshot`](#f3snapshot) bytes as the second data block in the CAR when `F3Data != nil` in the metadata.

### SnapshotMetadata

```go

> type SnapshotMetadata {

Reviewer: Why don't we have a version field?

Reviewer 2: Not critical though, versioning can just be done by breaking the format and then checking schemas -- does this thing decode as a V1 format? No, then try V2. Just as long as they are different enough, like having a different number of fields.

    HeadTipsetKey []Cid // required
    F3Data        *Cid  // optional
}
```

### F3Snapshot

An F3 snapshot contains one header block and N(N>0) data blocks in the below format:

`[Header block] [Data block] [Data block] [Data block] ...`

A header block is a CBOR-encoded [`F3SnapshotHeader`](#f3snapshotheader) with a length prefix in the below format:

`[varint-encoded length] [CBOR-encoded F3SnapshotHeader]`

### F3SnapshotHeader

```go
type SnapshotHeader struct {

> 	Version           uint64

Author: That's a great idea

	FirstInstance     uint64
	LatestInstance    uint64

> 	InitialPowerTable gpbft.PowerEntries

Reviewer: Do we define this structure anywhere?

}
```

A data block is a CBOR-encoded [`FinalityCertificate`](#finalitycertificate) with a length prefix in the below format:

`[varint-encoded length] [CBOR-encoded FinalityCertificate]`

### FinalityCertificate

```go
// FinalityCertificate represents a single finalized GPBFT instance.
type FinalityCertificate struct {
	// The GPBFT instance to which this finality certificate corresponds.
	GPBFTInstance uint64
	// The ECChain finalized during this instance, starting with the last tipset finalized in
	// the previous instance.
	ECChain *gpbft.ECChain
	// Additional data signed by the participants in this instance. Currently used to certify
	// the power table used in the next instance.

> 	SupplementalData gpbft.SupplementalData

Reviewer 2: worth defining or linking

	// Indexes in the base power table of the certifiers (bitset)
	Signers bitfield.BitField
	// Aggregated signature of the certifiers
	Signature []byte
	// Changes between the power table used to validate this finality certificate and the power
	// used to validate the next finality certificate. Sorted by ParticipantID, ascending.

> 	PowerTableDelta PowerTableDiff `json:"PowerTableDelta,omitempty"`

Reviewer 2: worth defining, I think https://github.com/filecoin-project/go-f3/blob/d713c5a5ec94f0e6af7736f5c0b4bae625496acf/certs/certs.go#L18-L31 is relevant

Author: Added to the FRC

}
```

Notes:
- `FinalityCertificate`s should be ordered by `GPBFTInstance` in ascending order, thus they can be validated and intermediate power tables can be generated while data blocks are being read in a stream.
- The first and last `FinalityCertificate` instances should match those in the header, respectively.

## Backwards Compatibility
<!--All FIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The FIP must explain how the author proposes to deal with these incompatibilities. FIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->

> - A filecoin node should try to read a snapshot CAR in the proposed format, and fallback to the old format to maintain backward compatibility.

Reviewer 2: ```suggestion
- A Filecoin node should try to read a snapshot CAR in the proposed format. A failure to successfully decode the blocked referenced as as the CAR's single root using the schema presented above, a snapshot reader may fallback to the old format and maintain backward compatibility. Additional failures to decode original snapshot format would indicate a fatal error.
```
